Add unit test to verify DDP + RPC correctness.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40210 Add thrift and tensorpipe backend tests for test_ddp_under_dist_autograd.
* **#40139 Add unit test to verify DDP + RPC correctness.**

This unit test runs the same set of operations locally and then with
DDP + RPC to verify correctness.

Differential Revision: [D22084385](https://our.internmc.facebook.com/intern/diff/D22084385/)

pritamdamania committed:
Add unit test to verify DDP + RPC correctness.

This unit test runs the same set of operations locally and then with
DDP + RPC to verify correctness.

Differential Revision: [D22084385](https://our.internmc.facebook.com/intern/diff/D22084385/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add unit test to verify DDP + RPC correctness."


This unit test runs the same set of operations locally and then with
DDP + RPC to verify correctness.

Differential Revision: [D22084385](https://our.internmc.facebook.com/intern/diff/D22084385/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add unit test to verify DDP + RPC correctness."


This unit test runs the same set of operations locally and then with
DDP + RPC to verify correctness.

Differential Revision: [D22084385](https://our.internmc.facebook.com/intern/diff/D22084385/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_ddp_under_dist_autograd.py:
```diff
@@ -545,6 +546,51 @@ def test_ddp_dist_autograd_sparse_grads(self):
+        # Run local case.
+        layer2 = nn.Linear(5, 1)
+        inputs = torch.rand((10, 10))
+        ddp_model = DistributedDataParallel(layer2)
```
Any reason why the ddp_model doesn't wrap the layer1/remote layer for both cases? 
i.e. `local_ddp_model = DDP(nn.Sequential(layer1, layer2))`, and `ddp_model_with_remote = DDP(nn.Sequential(remote_layer_1, layer2))`? Should the test still work as expected in this case?

pritamdamania87 replied:
In this test, I didn't intend to replicate layer1 via DDP thats why its separate. Also, if we put remote_layer1 as part of DDP, it won't be replicated whereas the local layer1 would be replicated. So that might cause some discrepancy in the results.

rohan-varma reviewed (approved):
Looks good!

pritamdamania committed:
Update on "Add unit test to verify DDP + RPC correctness."


This unit test runs the same set of operations locally and then with
DDP + RPC to verify correctness.

Differential Revision: [D22084385](https://our.internmc.facebook.com/intern/diff/D22084385/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add unit test to verify DDP + RPC correctness."


This unit test runs the same set of operations locally and then with
DDP + RPC to verify correctness.

Differential Revision: [D22084385](https://our.internmc.facebook.com/intern/diff/D22084385/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Add unit test to verify DDP + RPC correctness."


This unit test runs the same set of operations locally and then with
DDP + RPC to verify correctness.

Differential Revision: [D22084385](https://our.internmc.facebook.com/intern/diff/D22084385/)

[ghstack-poisoned]
